File: android/app/src/androidTest/java/host/exp/exponent/utils/ExpoTestRunner.java
package host.exp.exponent.utils;

import android.support.test.internal.runner.junit4.AndroidJUnit4ClassRunner;
import android.support.test.internal.util.AndroidRunnerParams;

import org.json.JSONArray;
import org.json.JSONObject;
import org.junit.runner.Description;
import org.junit.runner.manipulation.Filter;
import org.junit.runner.manipulation.NoTestsRemainException;
import org.junit.runners.model.InitializationError;

import host.exp.exponent.annotations.ExpoAlwaysPassThroughFilter;
import host.exp.exponent.annotations.ExpoDevModeTest;
import host.exp.exponent.annotations.ExpoSdkVersionTest;
import host.exp.exponent.annotations.ExpoTestSuiteTest;

public class ExpoTestRunner extends AndroidJUnit4ClassRunner {

  public final static String TEST_TYPES_KEY = "includeTestTypes";
  public final static String TEST_SUITE_TEST_TYPE = "test-suite";
  public final static String DEV_MODE_TEST_TYPE = "dev-mode";
  public final static String SDK_VERSIONS_KEY = "includeSdkVersions";


  public static class ExpoTestFilter extends Filter {

    JSONObject testConfig;

    public ExpoTestFilter() {
      testConfig = TestConfig.get();
    }

    @Override
    public boolean shouldRun(Description description) {
      if (description.getAnnotation(ExpoAlwaysPassThroughFilter.class) != null) {
        return true;
      }

      JSONArray testTypes = testConfig.optJSONArray(TEST_TYPES_KEY);
      if (testTypes != null) {
        boolean foundTestType = false;
        for (int i = 0; i < testTypes.length(); i++) {
          if (testTypes.optString(i).equals(TEST_SUITE_TEST_TYPE) && description.getAnnotation(ExpoTestSuiteTest.class) != null) {
            foundTestType = true;
          } else if (testTypes.optString(i).equals(DEV_MODE_TEST_TYPE) && description.getAnnotation(ExpoDevModeTest.class) != null) {
            foundTestType = true;
          }
        }

        if (!foundTestType) {
          return false;
        }
      }

      JSONArray sdkVersions = testConfig.optJSONArray(SDK_VERSIONS_KEY);
      ExpoSdkVersionTest sdkVersionAnnotation = description.getAnnotation(ExpoSdkVersionTest.class);
      if (sdkVersions != null && sdkVersionAnnotation != null) {
        boolean foundSdkVersion = false;
        for (int i = 0; i < sdkVersions.length(); i++) {
          if (sdkVersionAnnotation.value().equals(sdkVersions.optString(i))) {
            foundSdkVersion = true;
          }
        }

        if (!foundSdkVersion) {
          return false;
        }
      }

      return true;
    }

    @Override
    public String describe() {
      return "Filters tests based on TEST_CONFIG env var";
    }
  }

  public ExpoTestRunner(Class<?> klass, AndroidRunnerParams runnerParams) throws InitializationError {
    super(klass, runnerParams);

    ExpoTestFilter filter = new ExpoTestFilter();
    try {
      filter.apply(this);
    } catch (NoTestsRemainException e) {
      throw new RuntimeException(e);
    }
  }
}

File: apps/expo-go/android/app/src/androidTest/java/host/exp/exponent/utils/ExpoTestRunner.kt
package host.exp.exponent.utils

import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner
import host.exp.exponent.annotations.ExpoAlwaysPassThroughFilter
import host.exp.exponent.annotations.ExpoDevModeTest
import host.exp.exponent.annotations.ExpoSdkVersionTest
import host.exp.exponent.annotations.ExpoTestSuiteTest
import org.json.JSONObject
import org.junit.runner.Description
import org.junit.runner.manipulation.Filter
import org.junit.runner.manipulation.NoTestsRemainException

class ExpoTestRunner(klass: Class<*>?) : AndroidJUnit4ClassRunner(klass) {
  class ExpoTestFilter : Filter() {
    private val testConfig: JSONObject = TestConfig.get()

    override fun shouldRun(description: Description): Boolean {
      if (description.getAnnotation(ExpoAlwaysPassThroughFilter::class.java) != null) {
        return true
      }
      val testTypes = testConfig.optJSONArray(TEST_TYPES_KEY)
      if (testTypes != null) {
        var foundTestType = false
        for (i in 0 until testTypes.length()) {
          if (testTypes.optString(i) == TEST_SUITE_TEST_TYPE && description.getAnnotation(
              ExpoTestSuiteTest::class.java
            ) != null
          ) {
            foundTestType = true
          } else if (testTypes.optString(i) == DEV_MODE_TEST_TYPE && description.getAnnotation(
              ExpoDevModeTest::class.java
            ) != null
          ) {
            foundTestType = true
          }
        }
        if (!foundTestType) {
          return false
        }
      }
      val sdkVersions = testConfig.optJSONArray(SDK_VERSIONS_KEY)
      val sdkVersionAnnotation = description.getAnnotation(
        ExpoSdkVersionTest::class.java
      )
      if (sdkVersions != null && sdkVersionAnnotation != null) {
        var foundSdkVersion = false
        for (i in 0 until sdkVersions.length()) {
          if (sdkVersionAnnotation.value == sdkVersions.optString(i)) {
            foundSdkVersion = true
          }
        }
        if (!foundSdkVersion) {
          return false
        }
      }
      return true
    }

    override fun describe(): String {
      return "Filters tests based on TEST_CONFIG env var"
    }
  }

  companion object {
    const val TEST_TYPES_KEY = "includeTestTypes"
    const val TEST_SUITE_TEST_TYPE = "test-suite"
    const val DEV_MODE_TEST_TYPE = "dev-mode"
    const val SDK_VERSIONS_KEY = "includeSdkVersions"
  }

  init {
    val filter = ExpoTestFilter()
    try {
      filter.apply(this)
    } catch (e: NoTestsRemainException) {
      throw RuntimeException(e)
    }
  }
}
